lib/ps.js

```diff
@@ -1,131 +1,103 @@
-var spawn = require('child_process').spawn
+'use strict'
+
 var os = require('os')
+var bin = require('./bin')
+
 var PLATFORM = os.platform()
 
-/**
-  * Get pid informations through ps command
-  * @param   {Number|Number[]}  pid
-  * @param   {Object}           options
-  * @return  {Function}         callback(err, stat)
-  * on os x skip headers with pcpu=,rss=
-  * on linux it could be --no-header
-  * on solaris 11 can't figure out a way to do this properly so...
-  */
-function ps (pid, options, done) {
-  if (!Array.isArray(pid)) {
-    pid = [pid]
+function parseTime (timestr, fraction) {
```

Comment:
Please take a look at my algorithm for that it's more concise, just add the fraction to it and you're good to go. Also it doesn't use any regex (should be faster):

```
/**
 * Get ps etime in seconds
 * POSIX format should be: [[dd-]hh:]mm:ss
 * on linux take "etimes" which is already the number of seconds since process started
 *
 * @param {String}
 * @return seconds
 */
function parseTime (time) {
  // Example output 7-22:43:36
  var tmp = time.split(':')

  if (~tmp[0].indexOf('-')) {
    var daysAndHours = tmp[0].split('-')
    tmp[0] = daysAndHours[1] // hours
    tmp.unshift(daysAndHours[0]) // days
  }

  var l = tmp.length
  var k = Math.pow(60, l - 1)
  time = 0

  for (var i = 0; i < l; i++) {
    time += k * parseInt(tmp[i], 10)
    k /= 60
  }

  return time
}
```

Reply:
Actually `pow` is one of the slowest operation you can use on a computer 😅just like `log`.
Anyway how do you want to handle the fraction?

Reply:
I'm not sure, I just benchmarked our two functions and mine looks faster:

```
# yours
ok ~314 μs (0 s + 313642 ns)

# mine
ok ~148 μs (0 s + 148158 ns)
```

Reply:
I've added fractional seconds:

```
function parseTimeDos (time) {
  // Example output 7-22:43:36
  var tmp = time.split(':')

  if (~tmp[0].indexOf('-')) {
    var daysAndHours = tmp[0].split('-')
    tmp[0] = daysAndHours[1] // hours
    tmp.unshift(daysAndHours[0]) // days
  }

  var l = tmp.length
  var k = Math.pow(60, l - 1)
  time = 0

  for (var i = 0; i < l; i++) {
    time += k * parseInt(tmp[i], 10)
    k /= 60
  }

  if (PLATFORM === 'darwin') {
    var frac = tmp[i - 1].split('.')
    time += parseInt(frac[1], 10) * 600
  }

  return time * 1000
}
```

```
# yours
ok ~310 μs (0 s + 310036 ns)

# mine
ok ~196 μs (0 s + 196210 ns)
```

Reply:
Cool! The power of benchmarking.
Put them in a loop and run them at least 1000 times if you can. I'm curious about the results.

Reply:
sure men, test data: 
```
var data = ['1-02:03:04.12', '1-01:02:03.12', '13:27.12', '01:02:03.12', '0:00']
```

1000 times * 4:

```

# one
ok ~7.48 ms (0 s + 7484277 ns)

# two
ok ~6.72 ms (0 s + 6715629 ns)

```

There's always a significant enough gap.

There is the bench code:

```
var PLATFORM = 'darwin'

function parseTime (timestr, fraction) {
  var time = 0
  var tpart = timestr.replace(/-|\./g, ':').split(':')
  var i = tpart.length - 1
  if (i >= 0 && fraction && PLATFORM === 'darwin') { // Fractions of second
    time += parseInt(tpart[i--], 10) * 600
  }
  if (i >= 0) { // Seconds
    time += parseInt(tpart[i--], 10) * 1000
  }
  if (i >= 0) { // Minutes
    time += parseInt(tpart[i--], 10) * 60000
  }
  if (i >= 0) { // Hours
    time += parseInt(tpart[i--], 10) * 3600000
  }
  if (i >= 0) { // Days
    time += parseInt(tpart[i--], 10) * 86400000
  }
  return time
}


/**
 * Get ps etime in seconds
 * POSIX format should be: [[dd-]hh:]mm:ss
 * on linux take "etimes" which is already the number of seconds since process started
 *
 * @param {String}
 * @return seconds
 */
function parseTimeDos (time) {
  // Example output 7-22:43:36
  var tmp = time.split(':')

  if (~tmp[0].indexOf('-')) {
    var daysAndHours = tmp[0].split('-')
    tmp[0] = daysAndHours[1] // hours
    tmp.unshift(daysAndHours[0]) // days
  }

  var l = tmp.length
  var k = Math.pow(60, l - 1)
  time = 0

  for (var i = 0; i < l; i++) {
    time += k * parseInt(tmp[i], 10)
    k /= 60
  }

  if (PLATFORM === 'darwin') {
    var frac = tmp[i - 1].split('.')
    time += parseInt(frac[1], 10) * 600
  }

  return time * 1000
}

var bench = require('nanobench')

var data = ['1-02:03:04.12', '1-01:02:03.12', '13:27.12', '01:02:03.12', '0:00']

bench('one', function (b) {
  b.start()

  for (let i = 0; i < 1000; i++) {
    data.forEach(function (d) {
      parseTime(d)
    })
  }

  b.end()
})

bench('two', function (b) {
  b.start()

  for (let i = 0; i < 1000; i++) {
    data.forEach(function (d) {
      parseTimeDos(d)
    })
  }

  b.end()
})
```

I run in with `node bench.js`

Reply:
you really don't like my function do you :laughing:

Reply:
I just want to understand why is faster 😅
Can you re-run the bench
replacing
`var tpart = timestr.replace(/-|\./g, ':').split(':')`
with
`var tpart = timestr.split(/-|:|\./)`

Thanks
If it still faster then we will adopt it